src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -98,18 +108,61 @@ public void setEnableTopQueries(final MetricType metricType, final boolean isCur
         boolean isTopNFeatureCurrentlyDisabled = !queryInsightsService.isTopNFeatureEnabled();
 
         if (isTopNFeatureCurrentlyDisabled) {
-            super.setEnabled(false);
             if (!isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndStopQueryInsights();
+                checkAndStopQueryInsights();
             }
         } else {
-            super.setEnabled(true);
             if (isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndRestartQueryInsights();
+                checkAndRestartQueryInsights();
             }
         }
     }
 
+    /**
+     * Set search query metrics enabled to enable collection of search query categorization metrics
+     * @param searchQueryMetricsEnabled boolean flag
+     */
+    public void setSearchQueryMetricsEnabled(boolean searchQueryMetricsEnabled) {
+        boolean oldSearchQueryMetricsEnabled = queryInsightsService.isSearchQueryMetricsFeatureEnabled();
+        this.queryInsightsService.enableSearchQueryMetricsFeature(searchQueryMetricsEnabled);
+        if (searchQueryMetricsEnabled) {
+            if (!oldSearchQueryMetricsEnabled) {
+                checkAndRestartQueryInsights();
+            }
+        } else {
+            if (oldSearchQueryMetricsEnabled) {
+                checkAndStopQueryInsights();
+            }
+        }
+    }
+
+    /**
+     * Stops query insights service if no features enabled
+     */
+    public void checkAndStopQueryInsights() {
+        if (!queryInsightsService.isAnyFeatureEnabled()) {
+            super.setEnabled(false);
+            queryInsightsService.stop();
+        }
+    }
+
+    /**
+     * Restarts query insights service if any feature enabled
+     */
+    public void checkAndRestartQueryInsights() {
+        if (queryInsightsService.isAnyFeatureEnabled()) {
+            super.setEnabled(true);
+            queryInsightsService.stop();
+            queryInsightsService.start();
+        }
+    }
+
+    private void updateSettingsForSearchQueryMetrics() {
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(SEARCH_QUERY_METRICS_ENABLED_SETTING, v -> setSearchQueryMetricsEnabled(v));
+        setSearchQueryMetricsEnabled(clusterService.getClusterSettings().get(SEARCH_QUERY_METRICS_ENABLED_SETTING));
+    }
```

Comment:
We should not add method for this as this logic for adding settings update consumer should only be invoked once and better done inline